compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -467,175 +468,239 @@ private void checkXMLNamespacePrefixes(List<BLangXMLElementFilter> filters) {
+            }
+        }
+        return symTable.noType;
+    }
 
-                if (memberTypes.stream().anyMatch(memType -> memType.tag == TypeTags.BYTE)) {
-                    return setLiteralValueAndGetType(literalExpr, symTable.byteType);
-                }
+    private BType getFiniteTypeMatchWithIntLiteral(BLangLiteral literalExpr, BFiniteType finiteType,
+                                                   Object literalValue) {
+        BType intLiteralType = getFiniteTypeMatchWithIntType(literalExpr, finiteType);
+        if (intLiteralType != symTable.noType) {
+            return intLiteralType;
+        }
+        int typeTag = getPreferredMemberTypeTag(finiteType);
+        for (int tag = TypeTags.FLOAT; tag <= TypeTags.DECIMAL; tag++) {
+            if (typeTag == tag && literalAssignableToFiniteType(literalExpr, finiteType, tag)) {
+                BType type = symTable.getTypeFromTag(tag);
+                setLiteralValueForFiniteType(literalExpr, type);
+                literalExpr.value = String.valueOf(literalValue);
+                return type;
+            }
+        }
+        return symTable.intType;
+    }
 
-                finiteType = getFiniteTypeWithValuesOfSingleType((BUnionType) expType, symTable.byteType);
-                if (finiteType != symTable.semanticError) {
-                    BType setType = setLiteralValueAndGetType(literalExpr, finiteType);
-                    if (literalExpr.isFiniteContext) {
-                        // i.e., a match was found for a finite type
-                        return setType;
+    private BType getTypeOfIntegerLiteral(BLangLiteral literalExpr, Object literalValue, BType expType) {
+        if (TypeTags.isIntegerTypeTag(expType.tag) || expType.tag == TypeTags.BYTE) {
+            return getIntLiteralType(expType, literalValue);
+        } else if (expType.tag == TypeTags.FLOAT) {
+            if (!types.validateFloatLiteral(literalExpr.pos, String.valueOf(literalValue))) {
+                resultType = symTable.semanticError;
+                return resultType;
+            }
+            literalExpr.value = ((Long) literalValue).doubleValue();
+            return symTable.floatType;
+        } else if (expType.tag == TypeTags.DECIMAL) {
+            literalExpr.value = String.valueOf(literalValue);
+            return symTable.decimalType;
+        } else if (expType.tag == TypeTags.FINITE) {
+            BFiniteType finiteType = (BFiniteType) expType;
+            return getFiniteTypeMatchWithIntLiteral(literalExpr, finiteType, literalValue);
+        } else if (expType.tag == TypeTags.UNION) {
+            Set<BType> memberTypes = ((BUnionType) expType).getMemberTypes();
+            for (BType memType : memberTypes) {
+                if (TypeTags.isIntegerTypeTag(memType.tag) || memType.tag == TypeTags.BYTE) {
+                    BType intLiteralType = getIntLiteralType(memType, literalValue);
+                    if (intLiteralType == memType) {
```

Comment:
Since you do TypeTags.isIntegerTypeTag(memType.tag) in top, can intLiteralType != memType ?

Reply:
Yes. because in the `getIntLiteralType` function we check whether  value match to the  range of the subtype.  
Example 
``` ballerina
int:Signed8|int:Unsigned32 i = 510;
```